Support for Connection API resource

t0mk opened the pull request:
This PR will add support for Connection API resource, fixes #228 

State of implemented API calls:

* [x] `POST ​/organizations​/{organization_id}​/connections` Request a new connection for the organization
* [x] `POST ​/projects​/{project_id}​/connections` Request a new connection for the project's organization
* [x] `GET ​/organizations​/{organization_id}​/connections` List organization connections
* [x] `GET ​/projects​/{project_id}​/connections` List project connections
* [x] `GET ​/connections​/{id}` Get connection
* [ ] `PUT ​/connections​/{id}` Update connection
* [x] `DELETE ​/connections​/{id}` Delete connection
* [x] `GET ​/connections​/{id}​/events` Retrieve connection events
* [x] `GET ​/connections​/{id}​/ports​/{id}​/events` Retrieve connection port events
* [x] `GET ​/virtual-circuit​/{id}​/events` Retrieve connection events
* [x] `GET ​/connections​/{connection_id}​/ports` List a connection's ports
* [x] `GET ​/connections​/{connection_id}​/ports​/{id}` Get a connection port
* [x] `GET ​/connections​/{connection_id}​/ports​/{port_id}​/virtual-circuits` List a connection port's virtual circuits
* [x] `POST ​/connections​/{connection_id}​/ports​/{port_id}​/virtual-circuits` Create a new Virtual Circuit
* [x] `GET ​/virtual-circuits​/{id}` Get a virtual circuit
* [x] `PUT ​/virtual-circuits​/{id}` Update a virtual circuit
* [x] `DELETE ​/virtual-circuits​/{id}` Delete a virtual circuit

Signed-off-by: Tomas Karasek <tom.to.the.k@gmail.com>

t0mk commented:
@displague I took a look on the Connections in the portal and adjusted the code.

I also added more functions. Now we miss the Updates (PUTs) of Connection and Virtual Circuit, and the creation (POST) of Virtual Circuit - I wasn't able to do it in the API so I didn't know what params can be updated.

I noticed that `GET ​/projects​/{project_id}​/connections` which should list project connections will also list connections for other projects in the organization. That's probably an API bug.

displague commented on connections.go:
```diff
@@ -0,0 +1,245 @@
+
+type virtualCircuitsRoot struct {
+	VirtualCircuits []ConnectionVirtualCircuit `json:"virtual_circuits"`
+	Meta            meta                       `json:"meta"`
```
Somewhat unrelated to this PR, but I wonder why `meta` is not exported.

https://github.com/packethost/packngo/blob/master/packngo.go#L38-L39

I can't find the linter rule or best practice (https://golang.org/doc/effective_go.html or https://github.com/golang/go/wiki/CodeReviewComments) that states this, but IIRC it is something like "exported fields in unexported types are difficult to use".

t0mk replied:
@displague Maybe the meta only makes sense in context of listing API request.

displague commented on connections.go:
```diff
@@ -0,0 +1,245 @@
+	Speed           string                     `json:"speed,omitempty"`
+	Organization    *Organization              `json:"organization,omitempty"`
+	VirtualCircuits []ConnectionVirtualCircuit `json:"virtual_circuits,omitempty"`
+	LinkStatus      string                     `json:"link_status,omitempty"`
```
Do we want to create an enum for LinkStatus too? (I didn't look into what values are possible)

We haven't done much of this to this point, but I'm not opposed to introducing enums in new resources (or backfilling these for known enums).

displague replied:
Same comment would apply to any Status field in this PR: https://github.com/packethost/packngo/pull/234/files#diff-e772df221260f0ad6138a54d148e7ec9af6cd92a8ff527a1a376171e1c27f83fR80

What do you think, @t0mk ?

t0mk replied:
@displague it defintely makes sense to have enum there, but I would be guessing the possible values.

I checked the api-docs json for the metadata for interconnections, and the status is there just as string without enumerated values specified.

If you know where to look in the EqMetal internal codebase, please let me know and I'll make it enum.

displague replied:
I dug up some possible values, but it would be best to have them included in the API spec or at least verified. (@jordan0day)

jordan0day replied:
Good call, I will add a ticket to get these values added to the api spec

jordan0day commented:
> I noticed that `GET ​/projects​/{project_id}​/connections` which should list project connections will also list connections for other projects in the organization. That's probably an API bug.

This is actually the intended result, though I agree it might seem a little odd.

Connections do not "belong to" projects -- they "belong to" organizations. The individual virtual circuits running on top of the connections do "belong to" projects, though. Especially in the case of *dedicated* connections, an organization might have many different projects, each with their own set of virtual circuits running on the same dedicated connection.

Because so many of our API routes are project-based, we felt it was pragmatic to go ahead and return the list of connections belonging to the organization here, so there would be less potential "why can't I see my connection?" confusion.

t0mk commented:
@jordan0day thanks for the info. I note that connections belong to organizations. I understand with the listing I guess then.

But does it then make sense to support POSTs to /project/<uuid>/connections to create connections? I did it in the code in the PR (there's Connections.OrganizationCreate and Connections.ProjectCreate). Seems that the ProjectCreate is unnecessary, even confusing.

t0mk commented:
@jordan0day, also, I wasn't able to test creating of Virtual circuits. Could I get a testing connection in project 52000fb2-ee46-4673-93a8-de2c2bdba33b? I have one request there, you can approve it, or create some more convenient connection for testing.

I want to try over the API and see what API requests it does and then I transcribe it in packngo.

jordan0day commented:
Hey @t0mk,

So in the API we do support creating a connection for the organization by just using the project ID, again for the same reason -- some people are used to just using projects for everything.

Personally I'd be okay with not having a `Connections.ProjectCreate`, or having `ProjectCreate` just call the same endpoint as `OrganizationCreate` by getting the organization ID from the project.

I'l DM you on the community slack about setting up a connection you can test with.

displague commented:
> Personally I'd be okay with not having a Connections.ProjectCreate, or having ProjectCreate just call the same endpoint as OrganizationCreate by getting the organization ID from the project.

packngo should align with whatever functions the Swagger spec presents. If a projectcreate is presented but serves no practical function, we could skip implementing it, but it sounds easy enough to implement and gets us better API parity to implement it.

We should note in the packngo Connectoions.ProjectList that the list is not project filtered, is identical to OrganizationList, and is offered for API parity.  

It would be great to see the API spec expand on the details around these endpoints.

t0mk commented:
@displague @jordan0day I fixed wrong types for some of the VLAN fields (str -> int), and I added code for attaching and removing VLANs to VirtualCircuits. I also moved VirtualCircuit handling to another source file.

Now waiting for testing Dedicated Connection so that I can test adding and removing Virtual Circuits.

t0mk commented:
@jordan0day thanks, I am testing creating VCs in the dedicated connection. When I create VC, I sometimes get a new resource created, but with Status "activation_failed". What are the most common causes of this? I get it when creating second VC in a Primary Port of a Dedicated Connection. The second VLAN is in the same project and facility as the first one.

jordan0day commented:
Sorry, that was my bad, I configured your dedicated connection wrong. Will fix and DM you on community slack.

t0mk commented:
@jordan0day thanks, now I can create 2 virtual circuits in one port from the code. It kind of all seems to work now. Except the `PUT ​/connections​/{id}` is still missing, I didn't see anything to update on the connection from the web Console.

@displague when you have time, would you take a look on the 2 new API services, connections.go and virtualcircuits.go; and tell what you don't like about it?

I was thinking 
* VirtualCircuits.RemoveVLAN to VirtualCircuits.UnassociateVLAN (to make it sound more like the web Console)
* VirtualCircuits.ConnectVLAN to VirtualCircuits.AssociateVLAN (to make it sound more like the previous bulletpoint)

displague commented:
@t0mk I think we should stick with `VirtualCircuitService.Update` for the following reasons:
* The name is predictable based on our other services
* This most closely follows the Swagger spec suggested name (and can be inferred from the action and URL)
* If another updateable field is introduced, we won't need to introduce an `Update` action that also Associate/Unassociate VLAN's in addition to the new update behavior.

In general, I think we should abide by the API names (which have compatibility assurances) and not the UI names (which offer not such guarantees).

t0mk commented:
@displague 
> I think we should stick with VirtualCircuitService.Update

done in ca4a578

t0mk commented:
.. will still fix the lint issues

displague commented:
Ready for a merge, @t0mk? (this is still in draft)

t0mk commented:
@displague The code is ready for review. I would feel better if someone else give it a try, but since the connections are so hard to set up for testing, we can merge. Just it would be good if we could mark it as experimental somehow.
Anyway, I will pay attention to bug reports for interconnection and will fix fast.

displague commented:
> Just it would be good if we could mark it as experimental somehow

I'll merge it now and we can start building the Terraform provider against this. That should help us find any rough edges in the experience.

We'll use our typical release tagging to indicate any changes that we need to make. Since we are still <1.0 we have more liberty to experiment.
